### PR TITLE
Java: Promote Unsafe URL Forward query from experimental

### PR DESCRIPTION
**Pull request by jcogs33**
This PR promotes `java/unsafe-url-forward-dispatch-load` from experimental (https://github.com/github/codeql/pull/6240 and https://github.com/github/codeql/pull/7286).

Note: this promotion does not include the resource-related additions in https://github.com/github/codeql/pull/8706 and [9199](https://github.com/github/codeql/pull/9199). Since we already model some of these resource methods ([example](https://github.com/github/codeql/blob/9c51514bd9fdd877990cb7a62a3688127002c392/java/ql/lib/ext/java.lang.model.yml#L6-L7)) as `path-injection` sinks, it probably makes more sense to add them to the `path-injection` query instead (see https://github.com/github/codeql/pull/15921). Let me know if you think any of these models should be added back to this query instead.

### CODE REVIEW

File: java/ql/lib/ext/javax.servlet.model.yml
Comment (jcogs33) [inline java/ql/lib/ext/javax.servlet.model.yml:18]: Let me know if it makes more sense to turn these models into summaries and have the sinks be `Arugment[this]` of [RequestDispatcher.forward](https://docs.oracle.com/javaee%2F7%2Fapi%2F%2F/javax/servlet/RequestDispatcher.html#forward-javax.servlet.ServletRequest-javax.servlet.ServletResponse-) and [RequestDispatcher.include](https://docs.oracle.com/javaee%2F7%2Fapi%2F%2F/javax/servlet/RequestDispatcher.html#include-javax.servlet.ServletRequest-javax.servlet.ServletResponse-) instead.

File: java/ql/lib/ext/experimental/io.undertow.server.handlers.resource.model.yml
Comment (atorralba) [inline java/ql/lib/ext/experimental/io.undertow.server.handlers.resource.model.yml:6]: Don't we want these models anymore?

File: java/ql/lib/ext/experimental/org.springframework.core.io.model.yml
Comment (atorralba) [inline java/ql/lib/ext/experimental/org.springframework.core.io.model.yml:16]: Same with these. Maybe they don't make sense anymore?

File: java/ql/lib/ext/javax.servlet.model.yml
Comment (atorralba) [reply java/ql/lib/ext/javax.servlet.model.yml:18]: I'm OK with the shorter paths, less chances we lose flow along the way 👍

File: java/ql/lib/semmle/code/java/security/PathSanitizer.qll
Comment (atorralba) [inline java/ql/lib/semmle/code/java/security/PathSanitizer.qll:71]: It's a bit weird that this is the only non-private sanitizer exposed by this library, but I understand why you did it. No need to change anything else, just taking mental note that it may make sense to expose all these sanitizers for consistency.

File: java/ql/lib/semmle/code/java/security/UrlForward.qll
Comment (atorralba) [inline java/ql/lib/semmle/code/java/security/UrlForward.qll:133]: The first line is implied by the second one.

```suggestion
  e = g.(CheckUrlEncodingGuard).getCheckedExpr() and
```

File: java/ql/lib/semmle/code/java/security/UrlForwardQuery.qll
Comment (atorralba) [inline java/ql/lib/semmle/code/java/security/UrlForwardQuery.qll:6]: Since we now use parameterized dataflow configurations, the assumption that `*Query.qll` files will only be imported from queries is no longer true. So let's not indirectly expose all these imports:

```suggestion
private import semmle.code.java.security.UrlForward
private import semmle.code.java.dataflow.FlowSources
private import semmle.code.java.security.PathSanitizer
```

(Actually, the `*Query.qll` file isn't necessary and this could be in `UrlForward.qll` directly, but I leave that up to you in case you still want to keep things separated).

File: java/ql/lib/semmle/code/java/security/UrlForwardQuery.qll
Comment (atorralba) [inline java/ql/lib/semmle/code/java/security/UrlForwardQuery.qll:22]: Small NIT if you want to avoid having to use parentheses:

```suggestion
    not exists(MethodCall mc, Method m |
      m instanceof HttpServletRequestGetRequestUriMethod or
      m instanceof HttpServletRequestGetRequestUrlMethod or
      m instanceof HttpServletRequestGetPathMethod
    |
      mc.getMethod() = m and
      mc = source.asExpr()
    )
```

File: java/ql/test/query-tests/security/CWE-552/UrlForwardTest.ql
Comment (atorralba) [inline java/ql/test/query-tests/security/CWE-552/UrlForwardTest.ql:18]: If you want to simplify:

```suggestion
import TestUtilities.InlineFlowTest
import semmle.code.java.security.UrlForwardQuery
import TaintFlowTest<UrlForwardQuery >
```

(but then you'll need to change the expectation message to `$ hasTaintFlow` in the test cases)

**Review by atorralba — COMMENTED** [review]
This looks great, good job.

I added a few comments but they are minor, so feel free to request the docs review before answering them.

File: java/ql/lib/ext/experimental/io.undertow.server.handlers.resource.model.yml
Comment (jcogs33) [reply java/ql/lib/ext/experimental/io.undertow.server.handlers.resource.model.yml:6]: I've added all resource-related models to a separate PR (https://github.com/github/codeql/pull/15921). Is it okay if I add them on that PR instead?

See my note from the PR description above:

> Note: this promotion does not include the resource-related additions in https://github.com/github/codeql/pull/8706 and https://github.com/github/codeql/pull/9199. Since we already model some of these resource methods ([example](https://github.com/github/codeql/blob/9c51514bd9fdd877990cb7a62a3688127002c392/java/ql/lib/ext/java.lang.model.yml#L6-L7)) as path-injection sinks, it probably makes more sense to add them to the path-injection query instead (see https://github.com/github/codeql/pull/15921). Let me know if you think any of these models should be added back to this query instead.

(sorry, re-reading my note, I did not make it clear that I had put resource-related taint steps as well as sinks in https://github.com/github/codeql/pull/15921).

File: java/ql/lib/ext/experimental/org.springframework.core.io.model.yml
Comment (jcogs33) [reply java/ql/lib/ext/experimental/org.springframework.core.io.model.yml:16]: Same as above for `ClassPathResource`.
`Resource.createRelative` was added as a `path-injection` sink in https://github.com/github/codeql/pull/14913.
A model for `ResourceLoader.getResource` [already exists](https://github.com/github/codeql/blob/8711232a388212846f71969b656f01f5269aecec/java/ql/lib/ext/org.springframework.core.io.model.yml#L7).

File: java/ql/lib/semmle/code/java/security/UrlForward.qll
Comment (jcogs33) [reply java/ql/lib/semmle/code/java/security/UrlForward.qll]: Oops, I missed this when cleaning up the code. Thanks for catching it.
Updated in https://github.com/github/codeql/pull/14854/commits/35fbc95cc743f63b16a68f8f5f6ce82a9d0bdbeb.

File: java/ql/lib/semmle/code/java/security/UrlForwardQuery.qll
Comment (jcogs33) [reply java/ql/lib/semmle/code/java/security/UrlForwardQuery.qll]: Updated in https://github.com/github/codeql/pull/14854/commits/121b24ea7c0e21f2032afc64b54f4de85bda9275.

File: java/ql/lib/ext/experimental/io.undertow.server.handlers.resource.model.yml
Comment (atorralba) [reply java/ql/lib/ext/experimental/io.undertow.server.handlers.resource.model.yml:6]: Ah, my bad, I hadn't read the description of this PR in a while. It's perfectly fine for this models to be in another PR 👍.

File: java/ql/lib/ext/experimental/org.springframework.core.io.model.yml
Comment (atorralba) [reply java/ql/lib/ext/experimental/org.springframework.core.io.model.yml:16]: It's fine as it is 👍

File: java/ql/test/query-tests/security/CWE-552/UrlForwardTest.ql
Comment (jcogs33) [reply java/ql/test/query-tests/security/CWE-552/UrlForwardTest.ql]: Updated in https://github.com/github/codeql/pull/14854/commits/2391fe7d89ff46cad715c09963c73ae6a8fd6f90.

File: java/ql/lib/semmle/code/java/security/UrlForwardQuery.qll
Comment (jcogs33) [reply java/ql/lib/semmle/code/java/security/UrlForwardQuery.qll]: Updated in https://github.com/github/codeql/pull/14854/commits/40c932a5f911c7ad5c1ca0cd5d517ee6081f4107.

File: java/ql/src/Security/CWE/CWE-552/UrlForward.qhelp
Comment (subatoi) [inline java/ql/src/Security/CWE/CWE-552/UrlForward.qhelp:16]:
```suggestion
<p>To guard against untrusted URL forwarding, you should avoid putting user input
directly into a forwarded URL. Instead, you should maintain a list of authorized
URLs on the server, then choose from that list based on the user input provided.</p>
```

Minor suggestions for sentence flow, second-person

File: java/ql/src/Security/CWE/CWE-552/UrlForward.ql
Comment (subatoi) [inline java/ql/src/Security/CWE/CWE-552/UrlForward.ql:3]:
```suggestion
 * @description URL forward based on unvalidated user input
```

V minor

**Review by subatoi — DISMISSED** [review]
Looks good! Only two minor comments

**Review by atorralba — DISMISSED** [review]
@jcogs33: leaving my technical approval, feel free to merge after @subatoi is happy with the docs.

**Comment by subatoi** [comment]
No need to wait for my re-approval if you're happy with my minor suggestions: FYI it's a UK 4-day weekend so I won't be back at work until the 2nd April after today.

**Comment by jcogs33** [comment]
Thanks @subatoi! I've applied your suggestions.

**Review by subatoi — APPROVED** [review]
Thank you!

**Pull request merged**
